**If you lost points on the last checkpoint you can get them back by responding to TA/IA feedback**  

Update/change the relevant sections where you lost those points, make sure you respond on GitHub Issues to your TA/IA to call their attention to the changes you made here.

Please update your Timeline... no battle plan survives contact with the enemy, so make sure we understand how your plans have changed.

# COGS 108 - Data Checkpoint

# Names

- Altair Aguelo
- Howard Hamilton
- Jason Tang
- Daniel Ramirez
- Ashton Bothun

# Research Question

-  Include a specific, clear data science question.
-  Make sure what you're measuring (variables) to answer the question is clear

This question should be specific, answerable with data, and clear. A general question with specific subquestions is permitted. (1-2 sentences). When we read this we should know clearly what your project will be about

What is the relationship between serve and match outcome in tennis(specifically point and match outcomes, aka points won within a set and match win/loss)? Does this reflect in the player's win percentage?
Some possible measures include first serve/second serve, placement of serve(wide, t, body), kind of point(regular, break), surface type, serve speed, etc. We will be looking specifically at men's pro tennis data to answer these questions.



## Background and Prior Work


- Include a general introduction to your topic
- Include explanation of what work has been done previously
- Include citations or links to previous work

This section will present the background and context of your topic and question in a few paragraphs. Include a general introduction to your topic and then describe what information you currently know about the topic after doing your initial research. Include references to other projects who have asked similar questions or approached similar problems. Explain what others have learned in their projects.

Find some relevant prior work, and reference those sources, summarizing what each did and what they learned. Even if you think you have a totally novel question, find the most similar prior work that you can and discuss how it relates to your project.

References can be research publications, but they need not be. Blogs, GitHub repositories, company websites, etc., are all viable references if they are relevant to your project. It must be clear which information comes from which references. (2-3 paragraphs, including at least 2 references)

 **Use inline citation through HTML footnotes to specify which references support which statements** 

For example: After government genocide in the 20th century, real birds were replaced with surveillance drones designed to look just like birds.<a name="cite_ref-1"></a>[<sup>1</sup>](#cite_note-1) Use a minimum of 2 or 3 citations, but we prefer more.<a name="cite_ref-2"></a>[<sup>2</sup>](#cite_note-2) You need enough to fully explain and back up important facts. 

Note that if you click a footnote number in the paragraph above it will transport you to the proper entry in the footnotes list below.  And if you click the ^ in the footnote entry, it will return you to the place in the main text where the footnote is made.

To understand the HTML here, `<a name="#..."> </a>` is a tag that allows you produce a named reference for a given location.  Markdown has the construciton `[text with hyperlink](#named reference)` that will produce a clickable link that transports you the named reference.

1. <a name="cite_note-1"></a> [^](#cite_ref-1) Lorenz, T. (9 Dec 2021) Birds Aren’t Real, or Are They? Inside a Gen Z Conspiracy Theory. *The New York Times*. https://www.nytimes.com/2021/12/09/technology/birds-arent-real-gen-z-misinformation.html 
2. <a name="cite_note-2"></a> [^](#cite_ref-2) Also refs should be important to the background, not some randomly chosen vaguely related stuff. Include a web link if possible in refs as above.


Our group is interested in the possible relationship between serve and match outcome in tennis. To conduct our analysis, we first want to be able to distinguish what makes a “good serve” by measuring aspects of the serve such as placement, speed, type, etc. If there is a measurable distinction, we hope to discover those variables and their significance to winning a point. Furthermore, through these measures we hope to see if the serve and match outcome relationship is reflected in a player’s win percentage, so we can see how important the serve is to a player’s game.

We are not the first people to analyze the tennis serve, and its effects on match outcomes; however, there are some differences in their work when compared to the analyses that we want to pursue. For example, Abhinav Maddineni, Samarth Gowda, Eric Liu, and Daniel Liang performed analyses on ATP data to find out what factors (if any) affect serves, if the type of court affects a match’s outcome, and what metrics are important for analyzing the competitiveness of a tennis game.<a name="cite_ref-1"></a>[<sup>1</sup>](#cite_note-1) Notably, their work focused on what factors affect the serve, like player height or court surface, and they made comparisons between winners and losers of matches by looking at various statistics. Our approach is fundamentally different because the data we have available looks at individual player statistics over the course of their entire career. Because the dataset is more granular, we hope to show how the serve affects individual success. Additionally, we seek to focus specifically on the serve and how it relates to a player's match outcomes and win percentage. 

In another article, we found that analyses have been performed on the probability of winning a point in men's elite tennis where the researchers found that the first serve is crucial to increase the probability of winning the point as well as the length of the rally.<a name="cite_ref-2"></a>[<sup>2</sup>](#cite_note-2) The main difference between our problem and the article’s is that the article has a general focus on what comes after the serve rather than a complete focus on the performance of the serve. We believe that through the similarities of incorporating different factors in a player's serve, this aspect of the article seems the most important in helping answer our research question. 

Another source was considering players from multiple different groups. This research has a similar idea of looking into the correlation between serves and the overall performance of a player.<a name="cite_ref-3"></a>[<sup>3</sup>](#cite_note-3)  We overlook statistics from different groups, since they compare data from all ages, skill level, and gender. While we only look at statistics from a few professional players. It is keen to note that they are looking to find the difference in serve amongst ages rather than individual success. They also noted that professionals had more success with their serves due to their experience and skills, whilst younger players struggled due to improper technique and strength.

The tennis serve has also been analyzed extensively from the perspective of biomechanics. Researchers have worked on determining the factors that affect tennis serve speed by analyzing the different motions that positively and negatively affect the serve speed for tennis players across age groups.<a name="cite_ref-4"></a>[<sup>4</sup>](#cite_note-4) Thus, based on their work, it seems clear that players can optimize their mechanics to improve their serve speed, and this may have an impact on their tennis game. This information is relevant to the professional players whose match data we aim to analyze because it is likely that their coaching teams use research to fine tune their serves to have every advantage possible on the court. We do not plan to analyze any data that pertains a player's swing mechanics, but it is important to keep this in mind.




1. <a name="cite_note-1"></a> [^](#cite_ref-1) Gowda, S., Liang, D., Liu, E., Maddineni A. (2021) Serving Up Visualizations: Examining Tennis Data. https://www.stat.cmu.edu/capstoneresearch/spring2021/315files/team13.html#
2. <a name="cite_ref-3"></a>[^](#cite_note-2) Prieto-Lage I. et al (28 Sep 2023) Match analysis and probability of winning a point in elite men's singles tennis. https://pmc.ncbi.nlm.nih.gov/articles/PMC10538650/?utm_source=chatgpt.com#sec013
3. <a name="cite_ref-3"></a>[^](#cite_note-3) Hizan H., Reid P., Whipp P. (Jan 2011) Comparison of serve an serve return statistics of high performance male and female tennis players from different age-groups. https://www.researchgate.net/publication/263213823_Comparison_of_serve_and_serve_return_statistics_of_high_performance_male_and_female_tennis_players_from_different_age-groups
4. <a name="cite_ref-4"></a> [^](#cite_note-4) Baiget, E., Brich Q., Colomar J., Corbi, F. (5 Jul 2022) Determinant Physical Factors of Tennis Serve Velocity: A Brief Review. https://doi.org/10.1123/ijspp.2022-0091

# Hypothesis



- Include your team's hypothesis
- Ensure that this hypothesis is clear to readers
- Explain why you think this will be the outcome (what was your thinking?)

What is your main hypothesis/predictions about what the answer to your question is? Briefly explain your thinking. (2-3 sentences)

We believe that there is likely to be a positive relationship between a strong serve and favorable match outcome because good serves can win a point outright or setup the momentum/aggression for the rest of the point. However, what makes a "good" serve is not concrete, and other variables and aspects of the serve (first serve/second serve, placement of serve(wide, t, body), kind of point(break, match), surface type, serve speed, etc.) are likely to affect match outcome. We hope to dictate what makes a "strong" serve and see if this reflects in win percentage.


# Data

## Data overview

Our data will be stored in csv files. This data was scraped and cleaned before being stored as a csv file.

For each dataset include the following information
- Dataset #1
  - Dataset Name:
  - Link to the dataset:
  - Number of observations:
  - Number of variables:
- Dataset #2 (if you have more than one!)
  - Dataset Name:
  - Link to the dataset:
  - Number of observations:
  - Number of variables:
- etc

Now write 2 - 5 sentences describing each dataset here. Include a short description of the important variables in the dataset; what the metrics and datatypes are, what concepts they may be proxies for. Include information about how you would need to wrangle/clean/preprocess the dataset

If you plan to use multiple datasets, add a few sentences about how you plan to combine these datasets.

The rows of each serve chart dataset pertains to matches of a player's career. Some important variables in this dataset include unreturned serves(unret%), points won in less than 3 shots after the serve (<=3W%), placement variables like D Wide%, and break point serves like BP Wide%. These variables allow us to explore different aspects of the serve and their relation to winning a point. The variables described are all stored as floats in our cleaned and tidied datasets, and are continuous, quantitative data. These variables could help us differentiate what makes a strong serve. For example, a high unreturned serve percentage is a good indicator that a serve is "strong" and effective. To wrangle the data, first we webscraped the data and gathered the information off of the [tennis abstract website](https://www.tennisabstract.com/). Cleaning the data involved getting rid of any "-" dashes found in the dataset, and converted it into nan. We then dropped any rows with these nans to clean the data. Furthermore, to tidy the data, we added columns to make the dataset more readable for future use. We separated the match column into year, tournament, and round. We simplified the result by splitting it into just win/loss as its own column, and the opponent into another column. We also changed the datatypes from strings into floats if the data was pertaining to numbers. Lastly, we added a column to represent the surface of the tournament venue for deeper analysis. To view our wrangling and cleaning process, please view the repository for the data cleaning notebooks. 

The rows of each serve speed dataset pertains to individual matches of a player's career as well. Some important variables in this dataset include avg speed, 1st serve avg speed, speed of specific serve placement types (t, wide), and max/min speed. These variables allow us to explore aspects of the speed of the serve, allowing us to further determine if speed is an important factor to the 'strength' of a serve and if it helps a player win their points/matches. The variables described are all also stored as floats in our cleaned in tidied datasets, and are also continuous, quantitative data since the data is numeric. These variables allow us to explore even deeper as to what makes a strong serve. For example, a high/consistent average speed could be a good indicator of a strong serve. If a serve is consistently fast, it should be harder to return, but that is something that we hope to explore further and solidify. To wrangle the data, we also webscraped the data and gathered the information from the [tennis abstract website](https://www.tennisabstract.com/). Cleaning the data once again involved getting rid of "-" dashes found, and converted these into nan. We also dropped any rows with these nans to clean the data. To tidy this data, we added columns to make the dataset more readable again. Again, we separated the match column into year, tournament, and round. We simplified the result column into win/loss and opponent into their own columns respectively. We again changed datatypes from strings to floats if the data pertained to numbers. Lastly we added a column to show the surface of the tournament venue, also hoping this would be useful for deeper analysis. For example, we could later check if serve speed somehow differed per surface type. To view our wrangling and cleaning process, the repository has the data cleaning notebooks available.

With these datasets, we could combine them into two separate datasets: one for serve chart and one for serve speed. We would just need to add a column representing who the player is at the beginning of the row. From there, we could concatenate the datasets/dataframes so that they would all be in one dataset. With these larger datasets made up of different players and their respective playstyles and serve strengths combined, we hope to have our datasets be stronger and better representative of the men's pro tennis population. 

To demo our data cleanliness and tidiness, we have provided the head and sum of nulls of each of the respective datasets below.


## Dataset #1: Djokovic Serve Chart
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=104925/Novak-Djokovic&table=mcp-serve) to dataset, csv also available for viewing in github.
- Number of observations: 492
- Number of variables: 23

In [55]:
import pandas as pd
djo = pd.read_csv('djokovic_serve_chart.csv')
djo.head(3)

year      tourney round result opponent  Unret%  <=3 W%  RiP W%  RiP W%.1  \
0  2025       Madrid   R64      L  Arnaldi    28.3    35.8    48.6      50.0   
1  2025  Monte Carlo   R32      L   Tabilo    14.3    23.2    50.0      56.5   
2  2025        Miami     F      L   Mensik    28.4    41.9    59.6      64.9   

   SvImpact  ...  RiP W%.3  SvImpact1st  D Wide%  A Wide%  BP Wide%  \
0      33.4  ...      50.0         53.6     38.5     45.0      50.0   
1      19.8  ...      56.5         33.7     56.3     42.9      25.0   
2      35.9  ...      64.9         43.9     57.7     66.7       0.0   

   2nd: Unret%  2nd: <=3 W%  2nd: RiP W%  2ndAgg  surface  
0          0.0          5.9         47.1    -118     clay  
1          4.2          4.2         43.5     -76     clay  
2         11.8         11.8         46.7       4     hard  

[3 rows x 23 columns]

In [56]:
print(djo.isnull().sum()) #no nulls in whole dataset!

year           0
tourney        0
round          0
result         0
opponent       0
Unret%         0
<=3 W%         0
RiP W%         0
RiP W%.1       0
SvImpact       0
1st: Unret%    0
1st: <=3 W%    0
RiP W%.2       0
RiP W%.3       0
SvImpact1st    0
D Wide%        0
A Wide%        0
BP Wide%       0
2nd: Unret%    0
2nd: <=3 W%    0
2nd: RiP W%    0
2ndAgg         0
surface        0
dtype: int64


## Dataset #2: Djokovic Serve Speed
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=104925/Novak-Djokovic&table=serve-speed) to dataset, csv also available for viewing in github.
- Number of observations: 100
- Number of variables: 19

In [61]:
import pandas as pd
djoSpeed = pd.read_csv('djokovic_serve_speed.csv')
djoSpeed.head(3)

year  tourney round result opponent  avg speed  1st avg  1st stdev  \
0  2024  US Open   R32      L  Popyrin      108.0    114.9        6.3   
1  2024  US Open   R64      W    Djere      106.0    117.9        4.9   
2  2024  US Open  R128      W    Albot      107.5    116.6        6.6   

   1st t avg  1st wide avg  max 1st  min 1st  2nd avg  2nd stdev  2nd t avg  \
0      117.7         112.9    124.0     94.0     94.8        9.5      104.6   
1      118.3         116.0    125.0    105.0     93.2       10.0      103.6   
2      119.9         110.8    126.0    104.0     96.7       11.0       94.8   

   2nd wide avg  max 2nd  min 2nd  
0          86.0    117.0     77.0  
1          97.6    119.0     78.0  
2         101.2    117.0     78.0

In [62]:
print(djoSpeed.isnull().sum())

year            0
tourney         0
round           0
result          0
opponent        0
avg speed       0
1st avg         0
1st stdev       0
1st t avg       0
1st wide avg    0
max 1st         0
min 1st         0
2nd avg         0
2nd stdev       0
2nd t avg       0
2nd wide avg    0
max 2nd         0
min 2nd         0
dtype: int64


## Dataset #3: Federer Serve Chart
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=103819/Roger-Federer&table=mcp-serve) to dataset, csv also available for viewing in github.
- Number of observations: 506
- Number of variables: 23

In [63]:
import pandas as pd
fed = pd.read_csv('federer_serve_chart.csv')
fed.head(3)

year    tourney round result opponent  Unret%  <=3 W%  RiP W%  RiP W%.1  \
0  2021  Wimbledon    QF      L  Hurkacz    22.6    43.0    43.5      52.4   
1  2021  Wimbledon   R16      W   Sonego    37.2    52.6    58.3      61.5   
2  2021  Wimbledon   R32      W   Norrie    38.6    56.1    52.9      56.5   

   SvImpact  ...  RiP W%.3  SvImpact1st  D Wide%  A Wide%  BP Wide%  \
0      34.1  ...      52.4         48.7     39.3     54.5      37.5   
1      44.0  ...      61.5         56.0     36.4     40.0     100.0   
2      48.8  ...      56.5         54.7     59.5     41.0       0.0   

   2nd: Unret%  2nd: <=3 W%  2nd: RiP W%  2ndAgg  surface  
0          6.9         27.6         29.6     -49    grass  
1         26.7         43.3         54.5     -30    grass  
2         36.8         47.4         45.8     -79    grass  

[3 rows x 23 columns]

In [32]:
print(fed.isnull().sum())

year           0
tourney        0
round          0
result         0
opponent       0
Unret%         0
<=3 W%         0
RiP W%         0
RiP W%.1       0
SvImpact       0
1st: Unret%    0
1st: <=3 W%    0
RiP W%.2       0
RiP W%.3       0
SvImpact1st    0
D Wide%        0
A Wide%        0
BP Wide%       0
2nd: Unret%    0
2nd: <=3 W%    0
2nd: RiP W%    0
2ndAgg         0
surface        0
dtype: int64


## Dataset #4: Federer Serve Speed
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=103819/Roger-Federer&table=serve-speed) to dataset, csv also available for viewing in github.
- Number of observations: 50
- Number of variables: 19

In [33]:
import pandas as pd
fedSpeed = pd.read_csv('federer_serve_speed.csv')
fedSpeed.head(3)

year    tourney round result opponent  avg speed  1st avg  1st stdev  \
0  2021  Wimbledon    QF      L  Hurkacz      107.0    112.7        6.5   
1  2021  Wimbledon   R16      W   Sonego      105.4    112.5        6.5   
2  2021  Wimbledon   R32      W   Norrie      106.8    112.5        7.6   

   1st t avg  1st wide avg  max 1st  min 1st  2nd avg  2nd stdev  2nd t avg  \
0      117.4         109.1    122.0     88.0     95.6        4.8       99.2   
1      116.5         106.4    122.0     94.0     94.3        7.0       97.5   
2      115.6         109.0    126.0     98.0     95.4        3.5       98.1   

   2nd wide avg  max 2nd  min 2nd surface  
0          95.4    102.0     85.0   grass  
1          92.8    106.0     79.0   grass  
2          93.3    102.0     87.0   grass

In [34]:
print(fedSpeed.isnull().sum())

year            0
tourney         0
round           0
result          0
opponent        0
avg speed       0
1st avg         0
1st stdev       0
1st t avg       0
1st wide avg    0
max 1st         0
min 1st         0
2nd avg         0
2nd stdev       0
2nd t avg       0
2nd wide avg    0
max 2nd         0
min 2nd         0
surface         0
dtype: int64


## Dataset #5: Isner Serve Chart
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=104545/John-Isner&table=mcp-serve) to dataset, csv also available for viewing in github.
- Number of observations: 82
- Number of variables: 23

In [43]:
import pandas as pd
#import numpy as np
isner = pd.read_csv('isner_serve_chart_clean.csv')
isner.head(3)



year  tourney round result opponent  Unret%  <=3 W%  RiP W%  RiP W%.1  \
0  2023  US Open   R64      L     Mmoh    48.1    63.0    55.7      64.3   
1  2023  Houston   R16      L  Brouwer    50.9    66.7    60.0      69.2   
2  2023   Dallas     F      L       Wu    64.4    71.2    41.5      32.0   

   SvImpact  ...  RiP W%.3  SvImpact1st  D Wide%  A Wide%  BP Wide%  \
0      55.6  ...      64.3         73.1     57.1     40.3       0.0   
1      58.2  ...      69.2         72.5     53.8     52.6       0.0   
2      67.0  ...      32.0         75.4     40.0     48.1      50.0   

   2nd: Unret%  2nd: <=3 W%  2nd: RiP W%  2ndAgg  surface  
0         19.6         34.8         45.9      69     hard  
1         45.5         59.1         50.0     103     clay  
2         42.9         53.6         56.3     -27     hard  

[3 rows x 23 columns]

In [44]:
print(isner.isnull().sum())

year           0
tourney        0
round          0
result         0
opponent       0
Unret%         0
<=3 W%         0
RiP W%         0
RiP W%.1       0
SvImpact       0
1st: Unret%    0
1st: <=3 W%    0
RiP W%.2       0
RiP W%.3       0
SvImpact1st    0
D Wide%        0
A Wide%        0
BP Wide%       0
2nd: Unret%    0
2nd: <=3 W%    0
2nd: RiP W%    0
2ndAgg         0
surface        0
dtype: int64


## Dataset #6: Isner Serve Speed
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=104545/John-Isner&table=serve-speed) to dataset, csv also available for viewing in github.
- Number of observations: 34
- Number of variables: 19

In [45]:
import pandas as pd
#import numpy as np
isnerSpeed = pd.read_csv('isner_serve_speed.csv')
isnerSpeed.head(3)



year    tourney round result opponent  avg speed  1st avg  1st stdev  \
0  2023    US Open   R64      L     Mmoh      115.6    119.8        8.6   
1  2023    US Open  R128      W     Diaz      117.1    120.3        5.4   
2  2022  Wimbledon   R32      L   Sinner      118.3    122.1        7.3   

   1st t avg  1st wide avg  max 1st  min 1st  2nd avg  2nd stdev  2nd t avg  \
0      122.4         119.2    134.0     68.0    105.1        5.1      104.8   
1      123.0         118.9    132.0    108.0    111.3        5.7      114.6   
2      126.5         119.5    136.0     95.0    109.9        6.0      115.0   

   2nd wide avg  max 2nd  min 2nd surface  
0         103.8    114.0     93.0    hard  
1         109.0    122.0     93.0    hard  
2         105.6    123.0     97.0   grass

In [46]:
print(isnerSpeed.isnull().sum())

year            0
tourney         0
round           0
result          0
opponent        0
avg speed       0
1st avg         0
1st stdev       0
1st t avg       0
1st wide avg    0
max 1st         0
min 1st         0
2nd avg         0
2nd stdev       0
2nd t avg       0
2nd wide avg    0
max 2nd         0
min 2nd         0
surface         0
dtype: int64


## Dataset #7: Medvedev Serve Chart
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=106421/Daniil-Medvedev&table=mcp-serve) to dataset, csv also available for viewing in github.
- Number of observations: 246
- Number of variables: 23

In [47]:
import pandas as pd
#import numpy as np
med = pd.read_csv('medvedev_serve_chart.csv')
med.head(3)



year       tourney round result opponent  Unret%  <=3 W%  RiP W%  RiP W%.1  \
0  2025   Monte Carlo   R16      L       De    11.5    13.5    30.2      30.4   
1  2025         Miami   R64      L    Munar    18.9    34.0    48.7      50.0   
2  2025  Indian Wells    SF      L     Rune    20.0    26.7    50.0      56.5   

   SvImpact  ...  RiP W%.3  SvImpact1st  D Wide%  A Wide%  BP Wide%  \
0      13.7  ...      30.4         24.5     40.0     21.4      33.3   
1      28.1  ...      50.0         38.4     75.0     60.0     100.0   
2      24.5  ...      56.5         32.3     46.7     20.0       0.0   

   2nd: Unret%  2nd: <=3 W%  2nd: RiP W%  2ndAgg  surface  
0          0.0          0.0         30.0    -118     clay  
1         16.7         33.3         46.7      32     hard  
2         16.7         20.0         44.0     -79     hard  

[3 rows x 23 columns]

In [48]:
print(med.isnull().sum())

year           0
tourney        0
round          0
result         0
opponent       0
Unret%         0
<=3 W%         0
RiP W%         0
RiP W%.1       0
SvImpact       0
1st: Unret%    0
1st: <=3 W%    0
RiP W%.2       0
RiP W%.3       0
SvImpact1st    0
D Wide%        0
A Wide%        0
BP Wide%       0
2nd: Unret%    0
2nd: <=3 W%    0
2nd: RiP W%    0
2ndAgg         0
surface        0
dtype: int64


## Dataset #8: Medvedev Serve Speed
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=106421/Daniil-Medvedev&table=serve-speed) to dataset, csv also available for viewing in github.
- Number of observations: 56
- Number of variables: 19

In [49]:
import pandas as pd
#import numpy as np
medSpeed = pd.read_csv('medvedev_serve_speed.csv')
medSpeed.head(3)



year  tourney round result opponent  avg speed  1st avg  1st stdev  \
0  2024  US Open    QF      L   Sinner      105.4    114.5        9.7   
1  2024  US Open   R16      W   Borges      111.2    118.8        7.5   
2  2024  US Open   R32      W  Cobolli      107.9    119.1        7.1   

   1st t avg  1st wide avg  max 1st  min 1st  2nd avg  2nd stdev  2nd t avg  \
0      119.3         109.8    126.0     92.0     92.2        6.3       91.0   
1      121.6         113.0    129.0    102.0     95.2        4.9      103.0   
2      122.2         115.3    127.0     96.0     90.7        6.9      103.0   

   2nd wide avg  max 2nd  min 2nd surface  
0          89.2    107.0     79.0    hard  
1          95.8    107.0     84.0    hard  
2          92.8    111.0     80.0    hard

In [50]:
print(medSpeed.isnull().sum())

year            0
tourney         0
round           0
result          0
opponent        0
avg speed       0
1st avg         0
1st stdev       0
1st t avg       0
1st wide avg    0
max 1st         0
min 1st         0
2nd avg         0
2nd stdev       0
2nd t avg       0
2nd wide avg    0
max 2nd         0
min 2nd         0
surface         0
dtype: int64


## Dataset #9: Monfils Serve Chart
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=104792/Gael-Monfils&table=mcp-serve) to dataset, csv also available for viewing in github.
- Number of observations: 129
- Number of variables: 23

In [51]:
import pandas as pd
#import numpy as np
mon = pd.read_csv('monfils_serve_chart.csv')
mon.head(3)



year       tourney round result  opponent  Unret%  <=3 W%  RiP W%  \
0  2025         Miami   R32      W     Munar    24.5    37.3    60.3   
1  2025         Miami  R128      W  Marozsan    34.9    50.6    54.9   
2  2025  Indian Wells   R32      L  Dimitrov    17.2    27.0    62.1   

   RiP W%.1  SvImpact  ...  RiP W%.3  SvImpact1st  D Wide%  A Wide%  BP Wide%  \
0      67.3      33.3  ...      67.3         46.0     45.7     47.4       0.0   
1      60.0      41.7  ...      60.0         54.8     43.3     58.3      33.3   
2      64.9      22.9  ...      64.9         32.3     50.0     47.2      50.0   

   2nd: Unret%  2nd: <=3 W%  2nd: RiP W%  2ndAgg  surface  
0          9.4         15.6         48.3     -29     hard  
1         19.2         30.8         47.6      72     hard  
2          9.5         16.7         57.9     -26     hard  

[3 rows x 23 columns]

In [52]:
print(mon.isnull().sum())

year           0
tourney        0
round          0
result         0
opponent       0
Unret%         0
<=3 W%         0
RiP W%         0
RiP W%.1       0
SvImpact       0
1st: Unret%    0
1st: <=3 W%    0
RiP W%.2       0
RiP W%.3       0
SvImpact1st    0
D Wide%        0
A Wide%        0
BP Wide%       0
2nd: Unret%    0
2nd: <=3 W%    0
2nd: RiP W%    0
2ndAgg         0
surface        0
dtype: int64


## Dataset #10: Monfils Serve Speed
- [Link](https://www.tennisabstract.com/cgi-bin/player-more.cgi?p=104792/Gael-Monfils&table=serve-speed) to dataset, csv also available for viewing in github.
- Number of observations: 47
- Number of variables: 19

In [53]:
import pandas as pd
#import numpy as np
monSpeed = pd.read_csv('monfils_serve_speed.csv')
monSpeed.head(3)



year    tourney round result     opponent  avg speed  1st avg  1st stdev  \
0  2024    US Open   R64      L         Ruud      107.7    115.5       11.2   
1  2024    US Open  R128      W  Schwartzman      103.9    111.2       11.4   
2  2024  Wimbledon   R32      L     Dimitrov      110.9    117.1        9.3   

   1st t avg  1st wide avg  max 1st  min 1st  2nd avg  2nd stdev  2nd t avg  \
0      116.8         110.5    133.0     94.0     97.7        7.1      101.5   
1      115.4         107.5    128.0     79.0     94.4        7.2       98.6   
2      120.8         111.5    131.0     98.0    100.2        7.6      108.3   

   2nd wide avg  max 2nd  min 2nd  
0         101.7    116.0     85.0  
1          95.8    107.0     80.0  
2         102.0    124.0     88.0

In [54]:
print(monSpeed.isnull().sum())

year            0
tourney         0
round           0
result          0
opponent        0
avg speed       0
1st avg         0
1st stdev       0
1st t avg       0
1st wide avg    0
max 1st         0
min 1st         0
2nd avg         0
2nd stdev       0
2nd t avg       0
2nd wide avg    0
max 2nd         0
min 2nd         0
dtype: int64


# Ethics & Privacy

- Thoughtful discussion of ethical concerns included
- Ethical concerns consider the whole data science process (question asked, data collected, data being used, the bias in data, analysis, post-analysis, etc.)
- How your group handled bias/ethical concerns clearly described

Acknowledge and address any ethics & privacy related issues of your question(s), proposed dataset(s), and/or analyses. Use the information provided in lecture to guide your group discussion and thinking. If you need further guidance, check out [Deon's Ethics Checklist](http://deon.drivendata.org/#data-science-ethics-checklist). In particular:

- Are there any biases/privacy/terms of use issues with the data you propsed?
  - proportion of data is biased to older/more popular players rather than new/not popular players, player bioinformation (hypothetically, what if we revealed an injury that was unknown because we measured their service was getting "weaker" or something like that), data might need to be scraped, data was crowdsourced (how trustworthy?)
- Are there potential biases in your dataset(s), in terms of who it composes, and how it was collected, that may be problematic in terms of it allowing for equitable analysis? (For example, does your data exclude particular populations, or is it likely to reflect particular human biases in a way that could be a problem?)
  - older/more popular players are more represented than new/not popular players, data collected by fans are inherently biased -> could introduce racial, cultural, gender bias, possible sabotage of data (how are changes to the dataset moderated)
- How will you set out to detect these specific biases before, during, and after/when communicating your analysis?
  - make sure datasets between players are around the same sample size
  - noting the era that the data was taken in
  - possibly cross examine the dataset with official ATP dataset
  - make sure to mention where these possible biases are found and how they might affect our analysis in our final communication
- Are there any other issues related to your topic area, data, and/or analyses that are potentially problematic in terms of data privacy and equitable impact?
  - player bioinformation (hypothetically, what if we revealed an injury that was unknown because we measured their service was getting "weaker" or something like that, would it affect the player's livelihood?)
  - players are represented unequally based on dataset, we get looks at more old/popular players over new/unpopular players
- How will you handle issues you identified?
  - ensure datasets have good sample size and are relatively equal in size, and possibly representative of real life expectation
  - note areas in our data that could be biased in our analysis and communicate this in our findings
  - cross examine the crowdsourced data with official data when we can

# Team Expectations 

Read over the [COGS108 Team Policies](https://github.com/COGS108/Projects/blob/master/COGS108_TeamPolicies.md) individually. Then, include your group’s expectations of one another for successful completion of your COGS108 project below. Discuss and agree on what all of your expectations are. Discuss how your team will communicate throughout the quarter and consider how you will communicate respectfully should conflicts arise. By including each member’s name above and by adding their name to the submission, you are indicating that you have read the COGS108 Team Policies, accept your team’s expectations below, and have every intention to fulfill them. These expectations are for your team’s use and benefit — they won’t be graded for their details.

* Communication through Discord
* Meetings through Discord, in person if needed
* Meet at least once weekly before assignments/checkpoints are due
* Commitment to task delegation: do what you said you're gonna do!
* If there's any issues/blockers, communicate that as soon as possible so that we can figure out what to do
  * ex if you got an exam, let someone know so there's time to help
* Friday/Weekend/once a week checkpoint/meeting, catch everyone up on what you're doing/did and maybe have work together time
  * make sure everyone knows whats going on, not a compartmentalized project
* Conflict resolution: talk to each other privately, don't blast in public chat
  * be civil, understand agree to disagree at times
* All of us meet at least once, maybe do something just to get to know each other
* Tell everyone when made changes to repo so that they can pull
* Pull before every time you make changes just to keep up to date
* If theres anything to work on, just communicate on Discord

# Project Timeline Proposal

Specify your team's specific project timeline. An example timeline has been provided. Changes the dates, times, names, and details to fit your group's plan.

If you think you will need any special resources or training outside what we have covered in COGS 108 to solve your problem, then your proposal should state these clearly. For example, if you have selected a problem that involves implementing multiple neural networks, please state this so we can make sure you know what you’re doing and so we can point you to resources you will need to implement your project. Note that you are not required to use outside methods.



| Meeting Date  | Meeting Time| Completed Before Meeting  | Discuss at Meeting |
|---|---|---|---|
| 4/29  |  7 PM | Read & Think about COGS 108 expectations; brainstorm topics/questions  | Determine best form of communication; Discuss and decide on final project topic; discuss hypothesis; begin background research | 
| 4/30  |  3 PM |  Do background research on topic | Discuss ideal dataset(s) and ethics; draft project proposal | 
| 4/30  | 9 PM  | Edit, finalize, and submit proposal; Search for datasets  | Discuss Wrangling and possible analytical approaches; Assign group members to lead each specific part   |
| 5/05  | 6 PM  | Import & Wrangle Data (everyone) | Review/Edit wrangling/EDA (everyone); Discuss Analysis Plan   |
| 5/09  | 6 PM  | Choose data/players to import/wrangle | Discuss proposal, update anything needed, start import/wrangle |
| 5/14  | 7 PM  | Finalize wrangling/import/clean/tidy| Complete data checkpoint, finalize it, reviewed blockers and cleaned up/tidied everything else, updates for points lost on issues |
| 6/08  | 6 PM  | Complete analysis; Draft results/conclusion/discussion (everyone)| Discuss/edit full project (group) |
| 6/11  | Before 11:59 PM  | NA | Turn in Final Project & Group Project Surveys |